### PR TITLE
feat(gatsby-source-contentful): batch download assets from Contentful

### PR DESCRIPTION
**Pull request by guerc**
## Description

At my company we're using Gatsby together with Contentful quite extensively. Downloading the assets (> 500) from Contentful for use in our own S3 bucket is quite important. Unfortunately, during dev we often encounter either of two errors on multiple machines and configurations:

- Downloading the assets gets stuck just before finishing (mostly around 97%, mostly macOS)
- Downloading them seems to succeed but may be just a timeout as most assets are missing judging by the infamous `'childImageSharp' of null` error (Linux).

Note that they do not occur in our production env (AWS Amplify) using an Amazon Linux docker container as build environment.

This is an attempt at fixing this by instead of downloading everything at once, the downloads are done in chunks (or batches) of 50 files. This way, I could resolve the above mentioned issues.

### CODE REVIEW

File: packages/gatsby-source-contentful/src/download-contentful-assets.js [hunk 1]
```diff
@@ -1,6 +1,8 @@
 const ProgressBar = require(`progress`)
 const { createRemoteFileNode } = require(`gatsby-source-filesystem`)
 
+const BATCH_SIZE = 50
```
Comment (muescha) [inline packages/gatsby-source-contentful/src/download-contentful-assets.js]: Maybe make this as an adjustable plugin option defaults to 50?

Comment (guerc) [reply packages/gatsby-source-contentful/src/download-contentful-assets.js]: Yes, this is perhaps a good idea. Just wanted to check if this fix is beneficial, first. Any thoughts?

**Comment by vladar** [comment]
Sounds like a good idea. Not sure what are the possible pitfalls. @pieh what do you think? I will keep asking you about Contentful until I finally acquire enough context on it 😄

**Comment by pieh** [comment]
This PR is more work around flaky `createRemoteFileNode` behaviour. We do open very many concurrent download requests. It works fine in CI environments as network there doesn't have as many bottlenecks as local env (just from description):
> Downloading the assets gets stuck just before finishing (mostly around 97%, mostly macOS)
> Downloading them seems to succeed but may be just a timeout as most assets are missing judging by the infamous 'childImageSharp' of null error (Linux).
> Note that they do not occur in our production env (AWS Amplify) using an Amazon Linux docker container as build environment.

So this is not strictly about contentful here (other then the fact that it init those download requests)

**Comment by guerc** [comment]
> This PR is more work around flaky `createRemoteFileNode` behaviour (utility from `gatsby-source-filesystem`). We do open very many concurrent download requests. It works fine in CI environments as network there doesn't have as many bottlenecks as local env (just from description):
> 
> > Downloading the assets gets stuck just before finishing (mostly around 97%, mostly macOS)
> > Downloading them seems to succeed but may be just a timeout as most assets are missing judging by the infamous 'childImageSharp' of null error (Linux).
> > Note that they do not occur in our production env (AWS Amplify) using an Amazon Linux docker container as build environment.
> 
> So this is not strictly about contentful here (other then the fact that it init those download requests)

Fair enough. So, the solution would be to fix `createRemoteFileNode` so that other source plugins querying remote sources would also benefit from it?

Anyway, perhaps this PR can at least help some people using Contentful for now or are there any drawbacks?

**Comment by pieh** [comment]
@guerc sorry for confusion, my comment was meant as a reply to @vladar . In no way I meant to imply that we shouldn't move forward with the PR. You put it nicely in

> Anyway, perhaps this PR can at least help some people using Contentful for now or are there any drawbacks?

**Comment by guerc** [comment]
> @guerc sorry for confusion, my comment was meant as a reply to @vladar . In no way I meant to imply that we shouldn't move forward with the PR. You put it nicely in
> 
> > Anyway, perhaps this PR can at least help some people using Contentful for now or are there any drawbacks?
> 
> The only possible drawback is that batching downloads like that will potentially negatively impact perf in environments that don't exhibit problems you described (not a given, just something to consider).

Okay, making it configurable (thus being able to disable it) might solve that problem.

**Comment by vladar** [comment]
@guerc Do you have time to continue this PR (make batching configurable)?

**Comment by pvdz** [comment]
@guerc I guess you're busy with other things. I understand that, it's been a long time since creating this PR. So I'm going to put this PR in draft mode (there are merge conflicts to resolve, at least). If you do decide to continue this PR you can fix the merge conflict and put it back into review mode. Thanks! :)

Also tagging @wardpeet as he might be interested in this PR when working on something related.

**Comment by me4502** [comment]
This can be closed as of dc0ce0c1152d486bc1e58ac7fcaac36924437de8

**Pull request closed without merging**